### PR TITLE
Remove external optimization

### PR DESCRIPTION
**Pull request by lastmjs**
This improves the initial Azle prerequisites installation by 5-7 minutes, removes ic-cdk-optimizer and ic-wasm, solves the cmake and cc issues, gets rid of rustfmt command line stuff, cuts the step 1 time in half on subsequent builds

Closes: https://github.com/demergent-labs/azle/issues/1012 and https://github.com/demergent-labs/azle/issues/223

### CODE REVIEW

File: src/compiler/typescript_to_rust/azle_generate/src/main.rs [hunk 1]
```diff
@@ -14,11 +16,31 @@ fn main() {
     let compiler_info_string = std::fs::read_to_string(compiler_info_path).unwrap();
     let compiler_info: CompilerInfo = serde_json::from_str(&compiler_info_string).unwrap();
 
+    let environment_variables = if let Some(envs) = args.get(2) {
```
Comment (lastmjs) [inline src/compiler/typescript_to_rust/azle_generate/src/main.rs:19]: Let's leave a comment here, consider reading the dfx.json file, remove as many command line arguments as possible

File: src/compiler/typescript_to_rust/azle_generate/src/main.rs [hunk 2]
```diff
@@ -14,11 +16,31 @@ fn main() {
...
+        main_js,
+        &compiler_info.plugins,
+        &environment_variables,
+    )
+    .to_string();
 
-    // TODO let's fix this now too
-    // TODO let's really do this
-    println!("{}", result);
+    let mut f = File::create("../src/lib.rs").expect("Unable to create file");
```
Comment (lastmjs) [inline src/compiler/typescript_to_rust/azle_generate/src/main.rs:43]: Can we use rustfmt as a library right here? Then we could get rid of the rustfmt command line thing in the Node.js code

File: src/compiler/typescript_to_rust/generate_rust_canister.ts [hunk 3]
```diff
@@ -24,61 +28,64 @@ export function generateRustCanister(
...
+            ...process.env,
+            CARGO_TARGET_DIR: GLOBAL_AZLE_TARGET_DIR,
+            CARGO_HOME: GLOBAL_AZLE_RUST_DIR,
+            RUSTUP_HOME: GLOBAL_AZLE_RUST_DIR
+        }
+    });
+
+    // TODO should we handle presenting errors to the user somehow here?
+    // TODO maybe we should tell them to just rerun with verbose if the status code is not 0
+    if (buildResult.status !== null && buildResult.status !== 0) {
```
Comment (lastmjs) [inline src/compiler/typescript_to_rust/generate_rust_canister.ts:74]: Make sure to handle error and status code...handle any possible failure within reason

File: src/compiler/typescript_to_rust/generate_rust_canister.ts [hunk 4]
```diff
@@ -24,61 +28,64 @@ export function generateRustCanister(
...
 
 function runAzleGenerate(
     compilerInfoPath: string,
     canisterPath: string,
-    rootPath: string
-): Result<Rust, AzleError> {
-    if (isVerboseMode()) {
-        console.info('running azle_generate');
+    canisterConfig: JSCanisterConfig
+) {
```
Comment (lastmjs) [inline src/compiler/typescript_to_rust/generate_rust_canister.ts:51]: Put the result back in

File: src/compiler/utils/types.ts [hunk 5]
```diff
@@ -19,8 +19,12 @@ export type JSCanisterConfig = Readonly<{
     ts: string;
     candid: string;
     wasm: string;
+    env: string[];
```
Comment (lastmjs) [inline src/compiler/utils/types.ts:22]: Should also be optional

File: src/compiler/typescript_to_rust/azle_generate/src/lib.rs [hunk 6]
```diff
@@ -22,8 +22,9 @@ pub fn generate_canister(
     ts_file_names: &Vec<String>,
     main_js: String,
     plugins: &Vec<Plugin>,
+    environment_variables: &Vec<(String, String)>,
 ) -> TokenStream {
-    TsAst::new(ts_file_names, main_js, plugins)
+    TsAst::new(ts_file_names, main_js, plugins, environment_variables)
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/typescript_to_rust/azle_generate/src/lib.rs:27]: Resolved

File: src/compiler/typescript_to_rust/azle_generate/src/main.rs (hunk 2 above)
Comment (lastmjs) [reply src/compiler/typescript_to_rust/azle_generate/src/main.rs:43]: Done, using a library instead

**Review by dansteren — APPROVED** [review]
It overall looks good. There are some polishes that could be done, different places/ways to handle errors and such but I'll tackle them in https://github.com/demergent-labs/azle/issues/1017

File: src/compiler/typescript_to_rust/generate_rust_canister.ts [hunk 7]
```diff
@@ -6,192 +6,96 @@ import {
...
     const compilerInfoPath = join(
         canisterPath,
-        rootPath,
+        canisterConfig.root,
         'azle_generate',
         'compiler_info.json'
     );
 
+    // TODO why not just write the dfx.json file here as well?
     writeFileSync(compilerInfoPath, JSON.stringify(compilerInfo));
```
Comment (dansteren) [inline src/compiler/typescript_to_rust/generate_rust_canister.ts:34]: Just had a thought. Just a mental viewpoint, not that we need to change the code. But it seems like we're writing a config file for azle_generate here. So maybe we call it `.azlegeneraterc`? Kinda makes sense to me. Then when you run azle_generate it always looks for that file so we don't have to pass a file name across the CLI.

This makes more sense to me than calling it `compiler_info.json` and having to specify it via the CLI each time.

Just something to think about.

File: src/compiler/install_rust_dependencies.ts [hunk 8]
```diff
@@ -1,16 +1,57 @@
...
+    if (isWasm32TargetInstalled()) {
+        return;
+    }
 
-    execSync(`"${installRustDependenciesPath}" ${azleVersion} ${rustVersion}`, {
-        stdio: 'inherit'
+    time('[0/2] 🏗️  Preparing prerequisites...', 'inline', () => {
+        const installRustDependenciesPath = resolve(
+            __dirname,
+            '../../install_rust_dependencies.sh'
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/install_rust_dependencies.ts:22]: I've gone back and using join in many places that I wasn't, but sure please keep cleaning these up, let's just keep that in mind moving forward which I have been doing.

**Review by dansteren — COMMENTED** [review]
Just wanted to surface these thoughts I has while reviewing. Just ideas for later.

**Pull request merged**